source/common/http/conn_manager_impl.cc

```diff
@@ -470,6 +488,62 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
       connection_manager_.config_, *snapped_route_config_, connection_manager_.random_generator_,
       connection_manager_.runtime_, connection_manager_.local_info_);
 
+  if (!cached_route_.valid()) {
+    cached_route_.value(snapped_route_config_->route(*request_headers_, stream_id_));
+  }
+
+  // Check for WebSocket upgrade request if the route is valid, and supports WebSockets.
+  if ((protocol == Protocol::Http11) && cached_route_.value()) {
+    const Router::RouteEntry* route_entry = cached_route_.value()->routeEntry();
+    // Websocket route entries cannot be redirects.
+    bool websocket_upgrade_allowed =
+        (route_entry != nullptr) && route_entry->allowWebSocketUpgrade();
+    bool websocket_upgrade_requested = Utility::isWebSocketUpgradeRequest(*request_headers_);
+
+    if (websocket_upgrade_requested && websocket_upgrade_allowed) {
+      ENVOY_STREAM_LOG(debug, "found websocket connection. (end_stream={}):", *this, end_stream);
+
+      // A very kludgey way to get a handle over the stream
+      // decoder! WsHandlerImpl needs the stream decoder in order to
+      // properly terminate connections with HTTP error codes when
+      // there is an error establishing connection to upstream.
+      // TODO (rshriram) - can decoder_filters_ be empty? Is there a better way to get
+      // a handle over the stream decoder filter callback instead of ** ?
+      std::list<ActiveStreamDecoderFilterPtr>::iterator entry = decoder_filters_.begin();
+      if (entry != decoder_filters_.end()) {
+        connection_manager_.ws_connection_ =
+            std::unique_ptr<WebSocket::WsHandlerImpl>(new WebSocket::WsHandlerImpl(
```

Comment:
I just need a handle to encodeHeaders from the WebSocket Impl. May be make ActiveStream implement this callback interface and pass a reference to the WsHandlerCallbacks to WsHandlerImpl so that I don't have to do this kludgey conversion.
I cannot instantiate a WsHandlerImpl inside the callback, without having to pass the same parameters to the callback constructor. The reason is that when making upstream connection, WsHandler needs to pass on the request headers, and the route entry (to perform prefix/host rewrites.